wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,553 @@
+	if needIV && ivLength > 0 {
+		iv, err = uuid.GenerateRandomBytes(ivLength)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if err = c.client.EncryptInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(mechanism, iv)}, key); err != nil {
+		return nil, nil, fmt.Errorf("failed to pkcs11 EncryptInit: %s", err)
+	}
+	var ciphertext []byte
+	if ciphertext, err = c.client.Encrypt(session, plaintext); err != nil {
+		return nil, nil, fmt.Errorf("failed to pkcs11 Encrypt: %s", err)
+	}
+
+	if iv != nil {
+		return append(iv, ciphertext...), &keyId, nil
+	} else {
+		return ciphertext, &keyId, nil
+	}
+}
+
+func (c *Pkcs11Client) Decrypt(ciphertext []byte, keyId *Pkcs11Key) ([]byte, error) {
+	session, err := c.GetSession()
+	if err != nil {
+		return nil, err
+	}
+	defer c.CloseSession(session)
+
+	if keyId == nil {
+		keyId = &Pkcs11Key{ label: c.keyLabel, id: c.keyId }
+	}
+
+	obj, err := c.FindKey(session, *keyId, pkcs11.CKA_DECRYPT)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(obj) != 1 {
+		return nil, fmt.Errorf("expected 1 object, got %d", len(obj))
+	}
+	key := obj[0]
+
+	mechanism, err := c.GetKeyMechanism(session, key)
+	if err != nil {
+		return nil, err
+	}
+
+	var iv []byte
+	needIV, ivLength := IsIvNeeded(mechanism)
+	if needIV && ivLength > 0 {
+		if len(ciphertext) < ivLength {
+			return nil, fmt.Errorf("encrypted DEK is too short")
+		}
+
+		iv = ciphertext[:ivLength]
+		ciphertext = ciphertext[ivLength:]
+	}
+
+	if err = c.client.DecryptInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(mechanism, iv)}, key); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 DecryptInit: %s", err)
+	}
+
+	var decrypted []byte
+	if decrypted, err = c.client.Decrypt(session, ciphertext); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 Decrypt: %s", err)
+	}
+	return decrypted, nil
+}
+
+// Create a PKCS11 client for the configured module.
+func (c *Pkcs11Client) GetClient() (*pkcs11.Ctx, error) {
+	if c.client != nil {
+		return c.client, nil
+	}
+	c.client = pkcs11.New(c.lib)
+	err := c.client.Initialize()
+	if err != nil {
+		c.client = nil
+		return nil, fmt.Errorf("failed to initialize PKCS11: %w", err)
+	}
+	return c.client, nil
+}
+
+func (c *Pkcs11Client) GetSlotForLabel() (uint, error) {
+	if c.slot != 0 {
+		return c.slot, nil
+	}
+	if c.tokenLabel == "" {
+		return 0, fmt.Errorf("not token label configured")
+	}
+	slots, _ := c.client.GetSlotList(true)
+	for _, slot := range slots {
+		tokenInfo, err := c.client.GetTokenInfo(slot)
+		if err == nil && tokenInfo.Label == c.tokenLabel {
+			c.slot = slot
+			break
+		}
+	}
+	if c.slot == 0 {
+		return 0, fmt.Errorf("failed to find token with label: %s", c.tokenLabel)
+	}
+	return c.slot, nil
+}
+
+// Open a session and perform the authentication process.
+func (c *Pkcs11Client) GetSession() (pkcs11.SessionHandle, error) {
+	if c.client == nil {
+		return 0, fmt.Errorf("PKCS11 not initialized")
+	}
+
+	if c.slot == 0 {
```

Comment:
@glatigny Ah, another bug. SoftHSM's default initialization suggests slot 0, so I'd suggest trying it first rather than immediately going to `GetSlotForLabel()`:

```
$ softhsm2-util --init-token --slot 0 --label "My token 1"
```

per https://wiki.opendnssec.org/softhsm2/

Reply:
My bad, I forgot SoftHSM assigns random IDs in that case:

```
Available slots:
Slot 1513794117
    Slot info:
        Description:      SoftHSM slot ID 0x5a3aaa45                                      
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:    f521fa3b5a3aaa45
        Initialized:      yes
        User PIN init.:   yes
        Label:            My token 1                      
Slot 1
    Slot info:
        Description:      SoftHSM slot ID 0x1                                             
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:                    
        Initialized:      no
        User PIN init.:   no
        Label:                                            
```